.circleci/config.yml

```diff
@@ -65,19 +65,24 @@ jobs:
           command: |
             shopt -s extglob
 
+            # Loop through unit tests
+            # Note: failed unit tests are skipped. Return the last failed unit
+            #   test exit code once the loop completes.
+            exit_code=0
             for demo in !($IGNORED_DIRS); do
               if [ -d "$demo" ]; then
-                (
-                  env="envs/$demo"
-                  . "$env/bin/activate"
+                env="envs/$demo"
+                . "$env/bin/activate"
 
-                  cd "$demo"
-                  echo "$demo"
-                  coverage run -m unittest discover
-                )
+                cd "$demo"
+                echo "$demo"
+                coverage run -m unittest discover || exit_code=$?
+                cd ..
```

Comment:
Two major issues with this code block.

1. Any command can fail, not just `coverage`. If that happens, your next test is *not* run, rather, the complete job fails. (That's because each job/command sequence is run with `bash -eo pipefail`. `-e` makes `bash` exit on the first command that returns non-zero status.)
&nbsp;
A **naive fix** would be to just revert to subshell and capture the exit status of the whole subshell: `( ... ) || exit_code=$?`. However, `||` operator suppresses the standard `-e` behavior for all but the last command in a `||` list. In this case that applies to the complete subshell. And you can't even turn it back on with `set -e` in the subshell.
&nbsp;
So, the actual **solution** to this problem is a little bit more involved. For inspiration, I can direct you to [this SO answer](https://stackoverflow.com/questions/45590344/catch-signal-in-bash-but-dont-finish-currently-running-command/45591296#45591296) (you guessed it right, it's mine).
&nbsp;
You'll need to run the test subshell in background, waiting on it to finish, and only then capturing the status. Background subshell will inherit parent shell's options, i.e. `-e`, thus exit on first error. But since `wait` can be interrupted we need a loop.
```
declare -i pid status=255
( <unittest> ) & pid=$!
while (( status > 128 )); do
    wait $pid || status=$?
done
```

2. Your `exit_code` is overwritten on each test run, *not accumulated*. If you have only 2 tests, and first succeeds, but second fails, the complete *job fails*. But if the first fails, and second succeeds, the complete *job succeeds*. I don't think you want that.

Reply:
Instead of just talking about it, I took a shoot at fixing those issues. :smiley:
Please check my commit below and update it as necessary.

Reply:
Hm, my updates work on Unix, but not on Linux. :confused: :thinking:

Reply:
For (2), I am NOT overwriting on each test run. Rather, I am overwriting _only_ when the first condition (`coverage ..`) fails. Hence, I will have the last failing status, which would be good enough to at least let me know that something wrong. (aka I have the other failing statuses printed out)

Reply:
FYI, I am not disputing (1)! Just trying to wrap my head around it and study your fix.

Reply:
To your point in (1), you said "Any command can fail, not just coverage. If that happens, your next test is not run, rather, the complete job fails". 

If any other command fails, wouldn't I want it to fail fast? As in, I'm okay with my unit tests failing and thus, I am happy to just skip over them. However, if it's something outside of my unit test, wouldn't that be a larger problem and shouldn't I just fail early?

Reply:
^Also, yes, I do realize that that's a round about way of saying, "it's not a bug, it's a feature...". However, PLEASE do let me know why that would be naive!

Reply:
Ad (1), I wasn't sure what are your failure requirements (or design goals), i.e. what constitutes an *error* vs *failure*. So you're now saying you want to fail on errors in bash fast, but lazily in unit tests? I thought you wanted to always try running all tests. (And not fail due to, perhaps, some environment/package error specific to only one test.)

If that's true, then feel free to ignore my comments/code.

Also, my "naive fix" note above is not naive because such a design might be naive, but because it doesn't work in Bash, due to reasons explained (all errors are effectively suppressed in the subshell by `||` operator in `( ... ) || cmd`, even if you explicitly try turning it back on with `set -e`).

Ad (2), you're right. Aggregate exit status is the last failing test's status.

Reply:
Thank you for fixing the code and for the explanation on how you did it! I just had a few more questions about the fix